File: src/WePrint/Models/JobModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Threading.Tasks;

namespace WePrint.Models
{
    public class JobModel : DbModel, IIdempotentDbModel
    {
        public BidModel AcceptedBid { get; set; }
        public string CustomerId { get; set; }
        public string MakerId { get; set; }
        public AddressModel Address { get; set; }
        public List<BidModel> Bids { get; set; }
        public List<FileModel> Files { get; set; }
        public JobStatus Status { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public PrinterType PrinterType { get; set; }
        public MaterialType MaterialType { get; set; }
        public MaterialColor MaterialColor { get; set; }
        public string Notes { get; set; }
        public DateTime BidClose { get; set; }
        public List<CommentModel> Comments { get; set; }
        public int IdempotencyKey { get; set; }

        public void ApplyChanges(JobUpdateModel update)
        {
            ReflectionHelper.CopyPropertiesTo(update, this);
        }
    }

    public class JobUpdateModel
    {
        public string Id { get; set; }
        public int IdempotencyKey { get; set; }

        // Everything past here is optional
        public AddressModel Address { get; set; }
        public JobStatus? Status { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public PrinterType? PrinterType { get; set; }
        public MaterialType? MaterialType { get; set; }
        public MaterialColor? MaterialColor { get; set; }
        public string Notes { get; set; }
        public DateTime? BidClose { get; set; }
    }
}

File: src/WePrint/ClientApp/src/models/JobModel.js
import AddressModel from './AddressModel'

export default class JobModel {
	constructor() {
		this.id = "";
		this.name = "";
		this.customer = "";
		this.maker = "";
		this.status = "";
		this.description = "";
		this.printerType = "";
		this.materialType = "";
		this.materialColor = "";
		this.notes = "";
		this.bidClose = "";
		this.address = null;
	}

	static AllPropsEqual(a, b) {
		return (
			a.id === b.id &&
			a.name === b.name &&
			a.customer === b.customer &&
			a.maker === b.maker &&
			a.status === b.status &&
			a.description === b.description &&
			a.printerType === b.printerType &&
			a.materialType === b.materialType &&
			a.materialColor === b.materialColor &&
			a.notes === b.notes &&
			a.bidClose === b.bidClose &&
			AddressModel.AllPropsEqual(a.address, b.address)
		);
	}
}
